Observation API for History Metadata Storage

grigoryk opened the pull request:
TODO: ~~db migration (drop&recreate moz_places_metadata)~~, ~~missing Swift bindings update~~

This PR gets us closer to https://github.com/mozilla/application-services/issues/4067 - removes an explicit metadata `guid`, switching to an observation API (vs a CRUD api that was in place originally).

### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - Consider running `automation/all_tests.sh` to execute these tests locally.
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

mhammond commented on components/places/ffi/src/lib.rs:
```diff
@@ -486,32 +486,18 @@ pub extern "C" fn places_get_latest_history_metadata_for_url(
-        let guid = SyncGuid::from_string(guid.into_string());
-        storage::history_metadata::update_metadata(conn, &guid, total_view_time)
+        let observation: HistoryMetadataObservation = prost::Message::decode(buffer)?;
+        storage::history_metadata::note_observation(conn, observation)?;
```
`storage::history::` already has a `apply_observation`, so maybe `storage::history_metadata::apply_observation` or similar here too? or maybe `apply_metadata_observation` if having the same name smells wrong.

grigoryk replied:
Sure, `apply_metadata_observation` works.

mhammond reviewed (commented):
I only had a quick look, but this is looking great and is exactly what I had in mind when talking about this - thanks! I'll dig deeper next week.

grigoryk commented:
@mhammond I've updated this PR a bit - generally cleaned things up, fixed up tests, but mainly introduced a split ("key" vs "observation") at the api boundary level and in the guts, as well. I've left the protobuf layer mainly as-is, since it doesn't really buy us much to introduce the type split there as well.

Splitting things into `key` and `observation` helped quite a bit conceptually at the higher levels, but also helped clean up the queries as well. Not happy with the name `key`, of course, but neither @csadilek nor myself could come up with something better. Open to suggestions, but it's also not a terrible name and works for us.

Need to update the Swift bindings before landing this, but wanted to officially tag you for a review first :)

grigoryk commented:
Corresponding A-C PR in https://github.com/mozilla-mobile/android-components/pull/10346

mhammond commented on components/places/src/storage/history_metadata.rs:
```diff
@@ -2,47 +2,209 @@
+    }
+}
+
+trait CanDebounce {
```
nit: you don't need a trait for this, because no where do you use a generic type - ie, IIUC, you could just impl that function in each type and it would work fine. I'm fine if you think it adds some value though, but it also does add some cognitive overhead

grigoryk replied:
Cleaned this up, no more trait!

mhammond commented on components/places/src/storage/history_metadata.rs:
```diff
@@ -402,395 +504,516 @@ mod tests {
         };
     }
 
+    #[macro_use]
```
fancy :)

mhammond reviewed (approved):
Just a few nits, and there might be some cleanups worthwhile after we feel the "experimentation" phase is over, but there's nothing to block moving forward on this,

mhammond commented:
(and both CI failures seem real and related to this patch)

grigoryk commented on components/places/src/storage/history_metadata.rs:
```diff
@@ -2,47 +2,209 @@
+}
+
+impl PlaceEntry {
+    fn from(url: &str, tx: &PlacesTransaction<'_>, title: Option<String>) -> Result<Self> {
```
Switched to `fetch`.

grigoryk commented on components/places/src/storage/history_metadata.rs:
```diff
@@ -2,47 +2,209 @@
 use lazy_static::lazy_static;
 
+impl HistoryMetadataObservation {
+    fn from(observation_with_title: HistoryMetadataObservationWithTitle) -> (Self, Option<String>) {
```
Removed `HistoryMetadataObservationWithTitle` entirely and removed this as well.

grigoryk commented:
All right, just waiting for a green CI run now. Should be ready to land.
